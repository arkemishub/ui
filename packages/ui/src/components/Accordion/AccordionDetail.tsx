import { Transition } from "@headlessui/react";
import { twMerge } from "tailwind-merge";
import { ReactNode } from "react";

export interface IAccordionDetailProps {
  expanded?: boolean;
  children: ReactNode;
  expandIcon?: ReactNode;
}

function AccordionDetail({ children, expanded }: IAccordionDetailProps) {
  return (
    <Transition
      show={expanded}
      enter="transition-all ease-in-out duration-200"
      enterFrom="opacity-0"
      enterTo="opacity-100"
      leave="transition-all ease-in-out duration-200"
      leaveFrom="opacity-100"
      leaveTo="opacity-0"
      data-testid="arke-accordion-detail"
      className={twMerge(
        "accordion__detail",
        !expanded ? "h-0 overflow-hidden" : "h-[40px]"
      )}
    >
      {children}
    </Transition>
  );
}

export default AccordionDetail;
