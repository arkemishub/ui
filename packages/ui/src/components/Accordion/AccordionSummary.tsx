/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import { IAccordionProps } from "./Accordion.types";
import { twMerge } from "tailwind-merge";
import { ReactNode } from "react";

interface IAccordionSummaryProps extends Omit<IAccordionProps, "expanded"> {
  index?: number;
  expandIcon: ReactNode;
  expanded?: boolean;
}

function AccordionSummary({
  children,
  expanded,
  expandIcon,
  onChange,
  index,
  disabled,
}: IAccordionSummaryProps) {
  return (
    <div
      className="accordion__summary"
      tabIndex={index}
      role="button"
      onClick={() => !disabled && onChange?.()}
      onKeyDown={() => !disabled && onChange?.()}
      data-testid="arke-accordion-summary"
    >
      <div className="accordion__summary__content">{children}</div>
      <div
        className={twMerge(
          "accordion__summary__icon",
          !disabled ? "cursor-pointer" : "cursor-default",
          expanded && "rotate-180"
        )}
        data-testid="arke-accordion-icon"
      >
        {onChange && expandIcon}
      </div>
    </div>
  );
}

export default AccordionSummary;
