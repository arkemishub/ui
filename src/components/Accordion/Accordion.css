/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

.accordion {
    @apply p-2
}

.accordion__summary {
    @apply flex
}

.accordion__summary__content {
    @apply flex-1
}

.accordion__summary__icon {
    @apply transition-transform duration-500
}

.accordion__detail {
    @apply relative
}

.alert--success {
    @apply bg-success text-success-contrast
}
