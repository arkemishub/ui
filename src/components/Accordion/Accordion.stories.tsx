/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { IAccordionProps } from "./Accordion.types";
import React, { useState } from "react";
import Accordion, { AccordionDetail, AccordionSummary } from "./Accordion";

export default {
  title: "Accordion",
  component: Accordion,
};

const ExpandIcon = () => (
  <svg
    width="24"
    height="24"
    viewBox="0 0 24 24"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M19 9L12.7071 15.2929C12.3166 15.6834 11.6834 15.6834 11.2929 15.2929L5 9"
      stroke="#141C24"
      strokeWidth="1.5"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
  </svg>
);

export const Default = (args: Partial<IAccordionProps>) => {
  const [expanded, setExpanded] = useState<boolean>(false);
  return (
    <div className="flex flex-col gap-4">
      <Accordion expanded={expanded} onChange={() => setExpanded((p) => !p)}>
        <AccordionSummary expandIcon={<ExpandIcon />}>
          <p>Accordion 1</p>
        </AccordionSummary>
        <AccordionDetail>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse
            malesuada lacus ex, sit amet blandit leo lobortis eget.
          </p>
        </AccordionDetail>
      </Accordion>
      <hr />
      <Accordion
        expanded={false}
        disabled
        onChange={() => setExpanded((p) => !p)}
      >
        <AccordionSummary expandIcon={<ExpandIcon />}>
          <p>Disabled Accordion 3</p>
        </AccordionSummary>
        <AccordionDetail>
          <p>
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Suspendisse
            malesuada lacus ex, sit amet blandit leo lobortis eget.
          </p>
        </AccordionDetail>
      </Accordion>
    </div>
  );
};
