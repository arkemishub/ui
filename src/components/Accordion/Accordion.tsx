/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import React, {
  Children,
  cloneElement,
  isValidElement,
  ReactElement,
  ReactNode,
  useMemo,
} from "react";
import {
  IAccordionDetailProps,
  IAccordionProps,
  IAccordionSummaryProps,
} from "./Accordion.types";
import { twMerge } from "tailwind-merge";
import { Transition } from "@headlessui/react";

export const AccordionSummary = ({
  children,
  expanded,
  expandIcon,
  onChange,
  index,
  disabled,
}: IAccordionSummaryProps) => (
  <div
    className="accordion__summary"
    tabIndex={index}
    role="button"
    onClick={() => !disabled && onChange?.()}
    onKeyDown={() => !disabled && onChange?.()}
    data-testid="arke-accordion-summary"
  >
    <div className="accordion__summary__content">{children}</div>
    <div
      className={twMerge(
        "accordion__summary__icon",
        !disabled ? "cursor-pointer" : "cursor-default",
        expanded && "rotate-180"
      )}
      data-testid="arke-accordion-icon"
    >
      {onChange && expandIcon}
    </div>
  </div>
);

export const AccordionDetail = ({
  children,
  expanded,
}: IAccordionDetailProps) => (
  <Transition
    show={expanded}
    enter="transition-all ease-in-out duration-200"
    enterFrom="opacity-0"
    enterTo="opacity-100"
    leave="transition-all ease-in-out duration-200"
    leaveFrom="opacity-100"
    leaveTo="opacity-0"
    data-testid="arke-accordion-detail"
    className={twMerge(
      "accordion__detail",
      !expanded ? "h-0 overflow-hidden" : "h-[40px]"
    )}
  >
    {children}
  </Transition>
);

function Accordion({
  expanded,
  disabled,
  onChange,
  children,
}: IAccordionProps) {
  const content = useMemo(() => {
    return Children.map(
      children,
      (
        child: ReactElement<IAccordionProps & { index: number }>,
        index: number
      ) => {
        if (isValidElement(child) && child?.type === AccordionSummary)
          return cloneElement(child, {
            expanded,
            onChange,
            index,
            disabled,
          });
        if (isValidElement(child) && child?.type === AccordionDetail)
          return cloneElement(child, { expanded });
      }
    );
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [children]);

  return (
    <div
      className={twMerge("accordion", disabled && "opacity-50")}
      data-testid="arke-accordion"
    >
      {content}
    </div>
  );
}

export default Accordion;
