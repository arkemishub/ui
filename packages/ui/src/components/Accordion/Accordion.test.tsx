/**
 * Copyright 2023 Arkemis S.r.l.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { render } from "@testing-library/react";
import Accordion from "./Accordion";
import userEvent from "@testing-library/user-event";

const ExpandIcon = () => (
  <svg
    width="24"
    height="24"
    viewBox="0 0 24 24"
    fill="none"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M19 9L12.7071 15.2929C12.3166 15.6834 11.6834 15.6834 11.2929 15.2929L5 9"
      stroke="#141C24"
      strokeWidth="1.5"
      strokeLinecap="round"
      strokeLinejoin="round"
    />
  </svg>
);

describe("Accordion", () => {
  test("should match snapshot", () => {
    const { asFragment } = render(
      <Accordion expanded={true}>
        <Accordion.Summary expandIcon={<ExpandIcon />}>
          Accordion title
        </Accordion.Summary>
        <Accordion.Detail>Accordion title</Accordion.Detail>
      </Accordion>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should call onInputChange when input changes", async () => {
    const onExpandChange = jest.fn();
    const { getByTestId } = render(
      <Accordion expanded={true} onChange={onExpandChange}>
        <Accordion.Summary expandIcon={<ExpandIcon />}>
          Accordion title
        </Accordion.Summary>
        <Accordion.Detail>Accordion title</Accordion.Detail>
      </Accordion>
    );

    await userEvent.click(getByTestId("arke-accordion-icon"));
    expect(onExpandChange).toHaveBeenCalled();
  });
});
