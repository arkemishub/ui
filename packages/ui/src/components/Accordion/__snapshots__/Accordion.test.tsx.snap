// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="accordion"
    data-testid="arke-accordion"
  >
    <div
      class="accordion__summary"
      data-testid="arke-accordion-summary"
      role="button"
      tabindex="0"
    >
      <div
        class="accordion__summary__content"
      >
        Accordion title
      </div>
      <div
        class="accordion__summary__icon cursor-pointer rotate-180"
        data-testid="arke-accordion-icon"
      />
    </div>
    <div
      class="accordion__detail h-fit"
      data-testid="arke-accordion-detail"
    >
      Accordion title
    </div>
  </div>
</DocumentFragment>
`;
